go/libkb/sig_chain.go

```diff
@@ -665,7 +699,81 @@ func (sc *SigChain) VerifySigsAndComputeKeys(ctx context.Context, eldest keybase
 	// the id_table.  See LoadUser in user.go and
 	// https://github.com/keybase/go/issues/43
 
-	return
+	return cached, len(links), nil
+}
+
+func reverseListOfChainLinks(arr []ChainLinks) {
+	for i, j := 0, len(arr)-1; i < j; i, j = i+1, j-1 {
+		arr[i], arr[j] = arr[j], arr[i]
+	}
+}
+
+// VerifySigsAndComputeKeys iterates over all potentially all incarnations of the user, trying to compute
+// multiple subchains. It returns (bool, error), where bool is true if the load hit the cache, and false othewise.
+func (sc *SigChain) VerifySigsAndComputeKeys(ctx context.Context, eldest keybase1.KID, ckf *ComputedKeyFamily, loadAllSubchains bool) (bool, error) {
+	// First consume the currently active sigchain.
+	cached, numLinksConsumed, err := sc.verifySigsAndComputeKeysCurrent(ctx, eldest, ckf)
+	if !loadAllSubchains || err != nil || ckf.kf == nil {
+		return cached, err
+	}
+
+	allCached := cached
+
+	// ignore error here, since it shouldn't kill the overall load if historical subchains don't run
+	// correctly.
+	cached, _ = sc.verifySigsAndComputeKeysHistorical(ctx, numLinksConsumed, *ckf.kf)
+	if !cached {
+		allCached = false
+	}
+	return allCached, nil
+}
+
+func (sc *SigChain) verifySigsAndComputeKeysHistorical(ctx context.Context, numLinksConsumed int, kf KeyFamily) (allCached bool, err error) {
+
+	defer sc.G().CTrace(ctx, "verifySigsAndComputeKeysHistorical", func() error { return err })()
+	allLinks := sc.chainLinks
+	var cached bool
+
+	var prevSubchains []ChainLinks
+	sc.G().Log.CDebugf(ctx, "Continuing chain iteration into history: %d links total, but only %d consumed", len(allLinks), numLinksConsumed)
+
+	for {
+		allLinks = allLinks[0 : len(allLinks)-numLinksConsumed]
```

Comment:
Might be simpler to just take `allLinks` as an argument? Then we could overwrite that directly instead of having to update both `allLinks` and `numLinksConsumed` at different places in the loop?